File: ios/HearPortReceiver/Sources/HearPortAtomics/AtomicUInt64.c
#include "HearPortAtomics.h"

uint64_t hearport_atomic_load_u64(const uint64_t *value) {
    return __atomic_load_n(value, __ATOMIC_RELAXED);
}

uint64_t hearport_atomic_fetch_add_u64(uint64_t *value, uint64_t amount) {
    return __atomic_fetch_add(value, amount, __ATOMIC_RELAXED);
}

uint64_t hearport_atomic_exchange_u64(uint64_t *value, uint64_t replacement) {
    return __atomic_exchange_n(value, replacement, __ATOMIC_RELAXED);
}

bool hearport_atomic_compare_exchange_u64(uint64_t *value,
                                          uint64_t *expected,
                                          uint64_t replacement) {
    return __atomic_compare_exchange_n(value,
                                       expected,
                                       replacement,
                                       false,
                                       __ATOMIC_RELAXED,
                                       __ATOMIC_RELAXED);
}

File: ios/HearPortReceiver/Sources/HearPortReceiver/AtomicUInt64.swift
import Foundation
import HearPortAtomics

final class AtomicUInt64: @unchecked Sendable {
    private let storage: UnsafeMutablePointer<UInt64>

    init(_ initialValue: UInt64 = 0) {
        storage = .allocate(capacity: 1)
        storage.initialize(to: initialValue)
    }

    deinit {
        storage.deinitialize(count: 1)
        storage.deallocate()
    }

    func load() -> UInt64 {
        hearport_atomic_load_u64(storage)
    }

    @discardableResult
    func increment(by amount: UInt64 = 1) -> UInt64 {
        hearport_atomic_fetch_add_u64(storage, amount) + amount
    }

    @discardableResult
    func exchange(_ replacement: UInt64) -> UInt64 {
        hearport_atomic_exchange_u64(storage, replacement)
    }
}
